feat: insights-ansible-playbook-signer

m-horky opened the pull request:
* Card ID: CCT-691

Adds support for self-signed plays.
- Verifier now accepts hidden argument `--revocation-list`, which allows us to use custom revocation list instead of the one that's packaged.

Includes new setuptools entrypoint insights-ansible-playbook-signer.

Includes new executable in /usr/libexec.
- On Fedora and CentOS, this will later be a subpackage.
- On RHEL, we expect to only package the signing part.

The 'regression' CI job creates a playbook using the Signer and then hacks out the insights-client code to use custom GPG keys to verify it using the legacy/reference implementation on CentOS Stream 9.

ptoscano commented on .github/workflows/sanity.yml:
```diff
@@ -0,0 +1,76 @@
+      image: "quay.io/centos/centos:stream9"
+
+    steps:
+      - name: "Checkout the repository"
+        uses: actions/checkout@v4
```
run this after you install git-core (ie the step after this), otherwise there is the risk that the checked out repo is not a proper git repo

ptoscano commented on .github/workflows/sanity.yml:
```diff
@@ -0,0 +1,76 @@
+      # - sign own playbook
+      - name: "Prepare environment"
+        run: |
+          git clone https://github.com/RedHatInsights/insights-core.git --depth 1
```
simply use a github action for this:
```yaml
- uses: actions/checkout@v4
  with:
    repository: RedHatInsights/insights-core
    path: insights-core
```

m-horky replied:
I'd say this will slightly increase maintenance as next versions of the checkout action come out, but sure.

ptoscano commented on insights-ansible-playbook-verifier.spec:
```diff
@@ -26,17 +26,19 @@ playbooks generated by Red Hat Insights.
-%pyproject_install insights_ansible_playbook_lib insights_ansible_playbook_verifier
+%pyproject_install insights_ansible_playbook_lib insights_ansible_playbook_verifier insights_ansible_playbook_signer
 mkdir -p %{buildroot}/%{_libexecdir}
 mv %{buildroot}%{_bindir}/%{name} %{buildroot}%{_libexecdir}
+mv %{buildroot}%{_bindir}/insights-ansible-playbook-signer %{buildroot}%{_libexecdir}
 mkdir -p %{buildroot}/%{_localstatedir}/lib/%{name}/
-%pyproject_save_files insights_ansible_playbook_lib insights_ansible_playbook_verifier
+%pyproject_save_files insights_ansible_playbook_lib insights_ansible_playbook_verifier insights_ansible_playbook_signer
 
 %check
-%pyproject_check_import insights_ansible_playbook_lib insights_ansible_playbook_verifier
+%pyproject_check_import insights_ansible_playbook_lib insights_ansible_playbook_verifier insights_ansible_playbook_signer
 
 %files -n %{name} -f %{pyproject_files}
 %{_libexecdir}/%{name}
+%{_libexecdir}/insights-ansible-playbook-signer
```
hm I don't think the signer ought to be shipped with the verifier; it is mostly a developer tool

m-horky replied:
This was meant to be temporary. Once this patch was in, before we started packaging it into downstream distributions, we'd split this off into a separate package. I didn't want to do this in one giant step, the spec file may need further cleanups to be easier to maintain long-term.

ptoscano replied:
Then I'd not build the signer for now during RPM builds, and we can add it later as you correctly proposed.

ptoscano commented on python/insights_ansible_playbook_signer/app.py:
```diff
@@ -0,0 +1,179 @@
+        digest_file.write_bytes(play_digest)
+
+        subprocess.run(
+            ["rpm-sign", "--detachsign", "--key", key, "--nat", f"{digest_file!s}"],
```
ditto for `digest_file`

ptoscano commented on python/tests-integration/test_signer.py:
```diff
@@ -0,0 +1,140 @@
+    if pubkey.returncode != 0:
+        print(pubkey.stderr)
+        raise pytest.fail()
+    with open(f"{dir.name}/public.gpg", "w") as f:
+        f.write(pubkey.stdout)
```
ditto

ptoscano commented on python/tests-integration/test_signer.py:
```diff
@@ -0,0 +1,140 @@
+    if privkey.returncode != 0:
+        print(privkey.stderr)
+        raise pytest.fail()
+    with open(f"{dir.name}/private.gpg", "w") as f:
+        f.write(privkey.stdout)
```
ditto

ptoscano commented on python/tests-integration/test_signer.py:
```diff
@@ -0,0 +1,140 @@
+        text=True,
+        env={"GNUPGHOME": dir.name},
+    )
+    if genkey.returncode != 0:
+        print(genkey.stderr)
+        raise pytest.fail()
```
that's a big ugly TBH :)
there are already ways to log and raise cleanly for this kind of situations, so my suggestion is:
```python
print(genkey)
genkey.check_returncode()
```
ie unconditionally print the result of the process, so will be logged by pytest, and then raise in case of failure

ptoscano replied:
here `check_returncode()` was not applied, unlike done elsewhere, it seems

ptoscano replied:
:arrow_up:

ptoscano commented on python/tests-integration/test_signer.py:
```diff
@@ -0,0 +1,140 @@
+        text=True,
+        env={"GNUPGHOME": dir.name},
+    )
+    if pubkey.returncode != 0:
+        print(pubkey.stderr)
+        raise pytest.fail()
```
ditto

ptoscano commented on python/tests-integration/test_signer.py:
```diff
@@ -0,0 +1,140 @@
+        text=True,
+        env={"GNUPGHOME": dir.name},
+    )
+    if privkey.returncode != 0:
+        print(privkey.stderr)
+        raise pytest.fail()
```
ditto

ptoscano commented on python/tests-integration/test_signer.py:
```diff
@@ -0,0 +1,140 @@
+        check=False,
+        env={**os.environ, "LC_ALL": "C.UTF-8"},
+    )
+    if signing_result.returncode != 0:
+        print(signing_result.stderr.strip())
+    assert signing_result.returncode == 0
```
ditto

ptoscano commented on python/tests-integration/test_signer.py:
```diff
@@ -0,0 +1,140 @@
+        check=False,
+        env={**os.environ, "LC_ALL": "C.UTF-8"},
+    )
+    if reading_result.returncode != 0:
+        print(reading_result.stderr.strip())
```
ditto

and missing the check for `returncode == 0`?

m-horky replied:
That's strictly not necessary, but won't hurt.

ptoscano commented on python/tests-integration/test_signer.py:
```diff
@@ -0,0 +1,140 @@
+        [
+            "insights-ansible-playbook-signer",
+            "--playbook",
+            f"{DATA_DIRECTORY / playbook}",
```
no need for explicit stringification, `subprocess.run()` does it automatically

ptoscano commented on python/tests-integration/test_verifier.py:
```diff
@@ -37,9 +37,9 @@ def test_official_playbook(filename: str):
         stdout=subprocess.PIPE,
         stderr=subprocess.PIPE,
-        universal_newlines=True,
+        text=True,
```
unrelated change; please send it separately

ptoscano commented on python/tests-integration/test_verifier.py:
```diff
@@ -37,9 +37,9 @@ def test_official_playbook(filename: str):
+        text=True,
         check=False,
-        env={**os.environ, "LC_ALL": "UTF-8"},
+        env={**os.environ, "LC_ALL": "C.UTF-8"},
```
ditto

ptoscano commented on python/insights_ansible_playbook_signer/app.py:
```diff
@@ -0,0 +1,179 @@
+        # Ensure 'tasks' are the last element
+        if "tasks" in play.keys():
+            play["tasks"] = play.pop("tasks")
+        else:
+            logger.warning("Play does not include 'tasks', is that correct?")
```
I'd rather error out here, in case this situation is not wanted

m-horky replied:
If we error out there, we won't be able to sign the revocation list YAML. If it's too messy for you, I'd rather create a follow-up card.

ptoscano replied:
IMHO the signer ought to work in an atomic way: either it signs both the playbook and the revocation list, or none of them.

ptoscano reviewed (commented):
Some comments from a generic review (not a thorough one); will need to check the actual logic of the signer later on.

m-horky commented on python/insights_ansible_playbook_lib/crypto.py:
```diff
@@ -267,3 +267,35 @@ def verify_gpg_signed_file(
+        raise FileNotFoundError(f"Key '{key}' not found")
+
+    gpg = GPGCommand(
+        command=["--detach-sign", "--armor", f"{file.absolute()!s}"], key=key
```
I was following the "Explicit is better than implicit" rule. Sure though.

m-horky commented on python/insights_ansible_playbook_signer/app.py:
```diff
@@ -0,0 +1,179 @@
+
+        # Ensure 'tasks' are the last element
+        if "tasks" in play.keys():
+            play["tasks"] = play.pop("tasks")
```
See the comment above: this ensures the 'tasks' map will be the last element, and that it won't be somewhere in the middle between e.g. `hosts` and `vars`. While it is not required for the playbook to be parseable, it makes them much easier to validate when looking at them.

ptoscano commented on python/insights_ansible_playbook_signer/app.py:
```diff
@@ -0,0 +1,179 @@
+        if args.remote_key:
+            signature = send_signing_request(digest, key=args.remote_key)
+        else:
+            if not args.key.is_file():
+                raise RuntimeError(f"Key {args.key}' does not exist.")
```
In case of `send_signing_request()`, this check is done within the function; for consistence I'd do the same for `sign_play_digest()`, ie perform this existence check in `sign_play_digest()`.

ptoscano commented on python/insights_ansible_playbook_verifier/app.py:
```diff
@@ -68,17 +68,28 @@ def run() -> None:
+        logger.debug("Revocation digests obtained, can proceed to verification.")
+    else:
+        logger.warning("Revocation digest list loading skipped.")
+        digests = set()
```
minor nit: since `digest` is type-declared above:
```python
    digests: set[bytes]
```
then i'd simply initialize to an empty set there

m-horky commented:
I've addressed the code comments. The `.spec` file is still likely wrong, the Python RPM macros don't seem to easily support excluding a package, that will have to be fixed before merging.

ptoscano commented on python/insights_ansible_playbook_lib/__init__.py:
```diff
@@ -172,7 +173,7 @@ def verify_play(play: dict, gpg_key: bytes) -> bytes:
             )
             raise GPGValidationError(
-                "Play digest does not match its signature",
+                "Play digest does not match its signature.",
```
good change, although it's unrelated to the signer (and "pollutes" the history of this commit); what about a separate PR/commit?

ptoscano commented on python/insights_ansible_playbook_lib/crypto.py:
```diff
@@ -256,7 +256,7 @@ def verify_gpg_signed_file(
         )
 
-    gpg = GPGCommand(command=["--verify", f"{signature!s}", f"{file!s}"], key=key)
+    gpg = GPGCommand(command=["--verify", str(signature), str(file)], key=key)
```
ditto (good change but unrelated)

ptoscano commented on python/insights_ansible_playbook_verifier/app.py:
```diff
@@ -68,20 +68,36 @@ def run() -> None:
 
     # Load playbook with plays to verify
-    raw_playbook: str = ""
+    raw_playbook: str
```
ditto

ptoscano commented on python/insights_ansible_playbook_verifier/app.py:
```diff
@@ -103,7 +119,7 @@ def run() -> None:
         if digest in digests:
             raise RuntimeError(
-                f"Play '{play_name}'s digest is on revocation list: '{digest!s}'."
+                f"Digest of play '{play_name}' is on revocation list: '{bytearray(digest).hex()}'."
```
ditto

ptoscano commented on python/tests-unit/test_lib.py:
```diff
@@ -151,7 +133,7 @@ def test_missing_nested(self):
 
 
-class TestPlaybookDigest:
+class TestCreatePlayDigest:
```
ditto (good change but unrelated)

ptoscano commented on python/tests-unit/test_lib.py:
```diff
@@ -162,6 +144,38 @@ def test_ok(self, file: str):
         assert actual == expected
 
 
+class TestVerifyPlay:
+    @unittest.mock.patch(
+        "insights_ansible_playbook_lib.crypto.verify_gpg_signed_file",
+        return_value=unittest.mock.MagicMock(ok=False),
+    )
+    def test_requires_signature(self, _verify):
+        raw = {"name": "bad playbook", "tasks": [{"name": "a task"}]}
+
+        with pytest.raises(
+            lib.PreconditionError,
+            match="does not contain a signature",
+        ):
+            lib.verify_play(play=raw, gpg_key=b"")
+
+    @unittest.mock.patch(
+        "insights_ansible_playbook_lib.crypto.verify_gpg_signed_file",
+        return_value=unittest.mock.MagicMock(ok=False),
+    )
+    def test_requires_signature_exclude(self, _verify):
+        raw = {
+            "name": "bad playbook",
+            "vars": {"insights_signature": ""},
+            "tasks": [{"name": "a task"}],
+        }
+
+        with pytest.raises(
+            lib.PreconditionError,
+            match="does not have the key 'vars/insights_signature_exclude'",
+        ):
+            lib.verify_play(play=raw, gpg_key=b"")
+
+
```
ditto

ptoscano reviewed (changes requested):
It generally seems OK; few random notes:
- there are various changes that are good but unrelated to the addition of the signer; I recommend doing them in own PRs/commits, so they are easier to track (they are completely not documented in the commit message)
- if all the bits related to the signer are removed from `insights-ansible-playbook-verifier.spec`, then it will not be built & packaged together with the `insights-ansible-playbook-verifier`
  - this means that the integration tests for the signer should be skipped if the signer is not available (`shutil.which()`)
- this needs to be rebased to the current `main`

Thanks!

pkoprda commented on python/tests-integration/test_signer.py:
```diff
@@ -0,0 +1,166 @@
+    # Create fresh GPG keys
+    instructions_file = tmp_path / "instructions"
+    instructions_file.write_text(_GPG_INSTRUCTIONS)
+    genkey = subprocess.run(
+        [
+            "gpg",
+            "--batch",
```
Could you change the line `gpg` to `/usr/bin/gpg` so it would be consistent with explicit path usage from other files? Like it is already in `crypto.py`, `_keygen.py`, and `test_crypto.py`

pkoprda commented on python/tests-integration/test_signer.py:
```diff
@@ -0,0 +1,166 @@
+    gpg_public.write_text(pubkey.stdout)
+
+    # Save the private key
+    privkey = subprocess.run(
+        [
+            "gpg",
```
Ditto

pkoprda commented on python/tests-integration/test_signer.py:
```diff
@@ -0,0 +1,166 @@
+    genkey.check_returncode()
+
+    # Save the public key
+    pubkey = subprocess.run(
+        ["gpg", "--export", "--armor"],
```
Ditto

pkoprda reviewed (changes requested):
Looks good from my point of view, I just left one nitpick.
Also I agree with @ptoscano that unrelated changes should not be included in this PR.

ptoscano reviewed (approved):
[6 Months Later...](https://spongebob.fandom.com/wiki/List_of_time_cards)

LGTM, thanks!
